external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaOffsetReader.scala

```diff
@@ -419,6 +416,21 @@ private[kafka010] class KafkaOffsetReader(
     stopConsumer()
     _consumer = null  // will automatically get reinitialized again
   }
+
+  private def getPartitions(): ju.Set[TopicPartition] = {
+    consumer.poll(jt.Duration.ZERO)
+    var partitions = consumer.assignment()
+    val startTimeMs = System.currentTimeMillis()
+    while (partitions.isEmpty && System.currentTimeMillis() - startTimeMs < pollTimeoutMs) {
+      // Poll to get the latest assigned partitions
+      consumer.poll(jt.Duration.ofMillis(100))
```

Comment:
So using this new API will pull data to driver. Right? The previous `poll(0)` is basically a hack to avoid polling data in driver. Maybe we should ask the Kafka community to add a new API to pull metadata only.

Reply:
Good point. While Kafka doc says the behavior of such hack is indeterministic and Kafka never support it officially, we expect such behavior in any way.

I've initiated thread to ask about viable alternatives of `poll(0)` and possibility of adding public API to update metadata only.
https://lists.apache.org/thread.html/017cf631ef981ab1b494b1249be5c11d7edfe5f4867770a18188ebdc@%3Cdev.kafka.apache.org%3E

Reply:
I'm aware of that but since the doc says:
> there is no guarantee that poll(0) won't return records the first time it's called

I've considered `poll(0)` usage as design decision and no problem if small amount of data comes. Since you say it is not guaranteed but was working like that all the time the situation is different.

@HeartSaVioR thanks for initiating the discussion and let's see where it goes.

Reply:
Yeah you're right that's also true as well. One thing slightly different between two is, we are providing small amount of timeout (not 0) and we don't know how much amount of remaining timeout would be used as polling records (instead of polling metadata). It would be unlikely to 0 as it would be timed out if it goes below 0.